Add example for tone genesis

Huy-Ngo opened the pull request.

Huy-Ngo committed:
Start working on tonegen

Huy-Ngo committed:
Write apply_sin and enum for types

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,78 @@
+
+from palace import Buffer, Context, Device, Source
+
+CHUNK_LEN: int = 12000
+QUEUE_SIZE: int = 4
```
due to GH-10, we will not be able to play from decoder directly, i.e. we don't need this

McSinyx reviewed (commented):
The general idea is:
1. Create the tone decoder
2. Load to a buffer
3. Play it
4. ?
5. Profit

Huy-Ngo committed:
Write create_wave function

Huy-Ngo committed:
Write Decoder class for reading data from iterable

Pretty sure it doesn't work though.

Huy-Ngo committed:
Let the decoder read data and play it from buffer
Not running though

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,124 @@
+class DataDecoder(BaseDecoder):
+    def __init__(self, data: Iterable[float]):
+        self.data = data
+        pass
```
Don't store data here.  Something else to be cached though is the current frame number and the tone frequency.  Alternatively, we can even store the tone function as an attribute, i.e. `self.func = lambda frame: WAVEFORMS[waveform](frame/self.frequency, frequency)`.

Huy-Ngo replied:
Wait, how does this work? I just realize my solution doesn't make sense. `WAVEFORMS` now should be a list of functions, right?

McSinyx replied:
See the other comment.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,124 @@
+    def loop_points(self) -> Tuple[int, int]: return 0, 0
+
+    def read(self, count: int) -> bytes:
+        return float32(self.data[:count]).tobytes()
```
This should be able to create the data, i.e.
```python
from array import array # just realized that we don't really need numpy here
stop = min(self.start+count, self.length) # buffer is finite
data = array('f', map(self.func, range(self.start, stop)))
self.start = stop
return data.tobytes()
```

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,124 @@
+        for t in WAVE_TYPES:
+            print(t)
+        parser.exit()
+
+
+def apply_sin(data: Iterable[float], g: float,
+              srate: int, freq: int) -> None:
+    samps_per_cycle: float = srate / freq
+    for i in range(srate):
+        data[i] += sin((i/samps_per_cycle) % 1 * 2 * pi) * g
+
+
+def create_wave(wave_type: str, freq: int, srate: int) -> Iterable[float]:
+    data = [0] * srate
+    lim = int(srate/2/freq)
+    if wave_type == 'SINE':
+        apply_sin(data, pi, srate, freq)
+    elif wave_type == 'SQUARE':
+        for i in range(1, lim, 2):
+            apply_sin(data, 4 / pi * 1 / i, srate, freq * i)
+    elif wave_type == 'SAWTOOTH':
+        for i in range(1, lim):
+            apply_sin(data, 2 / pi * (1 if i % 2 else -1) / i, srate, freq * i)
+    elif wave_type == 'TRIANGLE':
+        for i in range(1, lim, 2):
+            apply_sin(data, 2 / pi * (1 if i / 2 % 2 else -1) / i,
+                      srate, freq * i)
+    elif wave_type == 'IMPULSE':
+        for i in range(srate):
+            data[i] = 0.0 if i % (srate/freq) else 1.0
+    elif 'WHITE_NOISE':
+        for i in range(srate):
+            data[i] += random() - random()
+    else:
+        print(f'{wave_type} not found, using the default value SINE:')
+        data = create_wave(wave_type, freq, srate)
+    return data
```
As mentioned, this conditional mess should be converted to a dictionary of string: `(time, sample_rate) -> value`.

Huy-Ngo replied:
When did you mention this? I can't recall that. Do you mean `WAVEFORMS` being the dictionary? Also, how can we map unknown non-constant like time and sample_rate in a dict?

Btw, there isn't `triangle` in `scipy.signal`, can I ignore that?

McSinyx replied:
> Do you mean WAVEFORMS being the dictionary?

Yes.

> Also, how can we map unknown non-constant like time and sample_rate in a dict?

We don't, e.g. `{'sine': lambda time, frequency: sin(time/frequency/pi/2)}`, or even better, `{'sine': sin}` and in ToneGenerator `self.func = lambda frame: WAVEFORMS[waveform](frame/self.frequency/frequency/pi/2)` (with `frequency` being the frequency of the sound like 420 Hz).

> Btw, there isn't triangle in scipy.signal, can I ignore that?

~~There is, just not under that name.  Look harder.~~  Just write a function for it then.

Huy-Ngo replied:
Ok, got it

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,124 @@
+
+class TypePrinter(Action):
+    def __call__(self, parser: ArgumentParser, *args, **kwargs) -> None:
+        print('Available waveform types:')
+        for t in WAVE_TYPES:
+            print(t)
```
`print` support varargs, i.e. `print('Available waveforms', *WAVEFORMS, sep='\n')`

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,124 @@
+                        help='print available waveform types in this example')
+    parser.add_argument('-w', '--waveform', default='SINE', nargs='+',
+                        help='audio files')
+    parser.add_argument('-d', '--device', default='', help='device name')
+    args = parser.parse_args()
+    play(args.device, args.waveform)
```
Device annotation is wrong (got string, expected Device).  Also we'll need the playback duration.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,124 @@
+            print('Generating waveform:', waveform)
+            while src.playing:
+                sleep(PERIOD)
+            print()
```
We don't need this.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,124 @@
+def play(device: Device, waveform: str) -> None:
+    with Device(device) as dev, Context(dev):
+        data: Iterable[float] = create_wave(waveform, 44100, 1)
+        dec = DataDecoder(data)
```
ToneGenerator might be a more intuitive name.

Huy-Ngo replied:
The decoder class?

McSinyx replied:
Yep.

Huy-Ngo committed:
Resolve some requests

- Edit the initializer
- Rename DataDecoder to ToneGenerator
- Edit ToneGenerator.read()
- Shorten printer action
- Fixing wrong typing for play()
- Remove unnecessary code

Huy-Ngo committed:
Fixing some linting issues and add argument

Huy-Ngo committed:
Add several signal waveforms

in the form as requested.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,127 @@
+
+
+class ToneGenerator(BaseDecoder):
+    def __init__(self, waveform: str):
+        self.func = lambda frame: WAVEFORMS[waveform](frame/self.frequency,
+                                           self.frequency)
```
As mentioned in https://github.com/McSinyx/palace/pull/69#discussion_r406212114, the wave frequency is not the sampling frequency.

Huy-Ngo replied:
I don't understand where does `frequency` comes from and thought that was a typo. Should I define that in the beginning of the file?

McSinyx replied:
Audio frequency should be parameterized.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,127 @@
+        parser.exit()
+
+
+def sine(time: int, sample_rate: int) -> Iterable[float]:
+    """Generate sinusoidal signal."""
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = sin(t)
+    return data
+
+
+def square(time: int, sample_rate: int) -> Iterable[float]:
+    """Generate square signal."""
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = signal.square(2 * pi * 5 * t)
+
+
+def sawtooth(time: int, sample_rate: int) -> Iterable[float]:
+    t = linspace(0, time, time * sample_rate, endpoint=False)
+    data = signal.sawtooth(2 * pi * 5 * t)
+
+
+def triangle(time: int, sample_rate: int) -> Iterable[float]:
+    pass
+
+
+def impulse(time: int, sample_rate: int) -> Iterable[float]:
+    pass
+
+
+def white_noise(time: int, sample_rate: int) -> Iterable[float]:
+    return random_sample((1, time * sample_rate))
```
These should be `(float) -> float` instead, i.e. for most you don't even have to write anything.

Huy-Ngo replied:
Hmm, isn't time and sample rate two different parameters? Btw, is it ok not to put the type hinting for WAVEFORMS, since when I put it flake8 says EOF in multiline or smt.

McSinyx replied:
Somehow I missed this ugh.

> Hmm, isn't time and sample rate two different parameters?

Well yes but actually no, since you'll divide them for each other anyway.

> since when I put it flake8 says EOF in multiline or smt

Could be a typo.

Huy-Ngo replied:
> Could be a typo.

True. I missed a `]` in all those clumsy brackets.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,127 @@
+
+from numpy import linspace
+from numpy.random import random_sample
+from scipy import signal
```
Absolute import please.

Huy-Ngo replied:
What do you mean? Isn't this absolute already?

McSinyx replied:
More absolute.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,127 @@
+def play(device: str, waveform: str, duration: int) -> None:
+    with Device(device) as dev, Context(dev):
+        dec = ToneGenerator(waveform)
+        _ = duration  # use it later
```
What do you thing that this will really do :smile:?  BTW for the tone generator to be defined, there needs to be 3 parameters:
1. waveform
2. frequency, i.e. pitch
3. duration, since we want it to be finite

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,127 @@
+    with Device(device) as dev, Context(dev):
+        dec = ToneGenerator(waveform)
+        _ = duration  # use it later
+        buf = Buffer.from_decoder(dec, 'tonegen')
+        buf.play()
```
Both Buffer and Source (returned by play) needs to be destroyed, i.e. use them as context managers.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,127 @@
+    parser = ArgumentParser()
+    parser.add_argument('-t', '--types', nargs=0, action=TypePrinter,
+                        help='print available waveform types in this example')
+    parser.add_argument('-w', '--waveform', default='SINE', nargs='+',
+                        help='audio files')
```
help string needs update, also it's fine if we only support playing one waveform at a time, for the sake of simplicity of default argument

Huy-Ngo replied:
Wait, I didn't remember copying from other files then not delete these params lol.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,127 @@
+    parser.add_argument('-w', '--waveform', default='SINE', nargs='+',
+                        help='audio files')
+    parser.add_argument('-d', '--device', default='', help='device name')
+    parser.add_argument('-l', '--duration', default=30, type=int,
+                        help='duration, in second')
```
Don't you think 30 second***s*** is a bit too long for the same sound to be played? Also use float to be more flexible.

Huy-Ngo committed:
Fix the signal generator functions and args

Huy-Ngo committed:
Finish initializer(?)

Huy-Ngo committed:
Handle buffer and source with context handler

Huy-Ngo committed:
Change arguments according to request

Huy-Ngo committed:
Remove wrong code

Huy-Ngo committed:
Fix linting

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,97 @@
+             'sawtooth': sawtooth,
+             'triangle': lambda time: return sawtooth(time, 0.5),
+             'impulse':unit_impulse,
+             'white noise': random}
```
`white-noise` might be a better choice to avoid user's need to escape the space.

Huy-Ngo committed:
Fix linting issues, fix syntax for lambda

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,97 @@
+
+class ToneGenerator(BaseDecoder):
+    def __init__(self, waveform: str, duration: float, frequency: int):
+        self.func = lambda frame: WAVEFORMS[waveform](frame/self.frequency,
+                                                      frequency)
```
Based on the signature, this should be `WAVEFORMS[waveform](frame/self.frequency/frequency)`.  Note that `sin` might need some tweak to have the period of 1.

Huy-Ngo replied:
Each of them needs some tweak I believe. But I don't know how.

McSinyx replied:
Finish the playback part so we can listen to it.

Huy-Ngo replied:
`buf.play` doesn't play it?

Anyway, I've got this and don't know what to do.

![image](https://user-images.githubusercontent.com/19296926/79090539-38274f00-7d74-11ea-808f-371f9a530e93.png)

McSinyx reviewed (changes requested):
So far so good.

Huy-Ngo committed:
Resolve request: white-noise, init

Huy-Ngo committed:
Add typing for dict

Huy-Ngo committed:
Set initial start attribute to 0

Huy-Ngo committed:
Correct the sample type

Error:
- `python3 palace-tonegen.py` runs, but doesn't give any sound
- If parameters are given, it raise an error: TypeError: unhashable type: 'list'

Huy-Ngo commented:
Still can't play it (it runs, but doesn't give sound). If I add any parameters, errors occur.

McSinyx commented:
> Still can't play it (it runs, but doesn't give sound).

Because you terminate before playback finishes.  Quickest hack is to `while src.playing: pass`.

> If I add any parameters, errors occur.

Please elaborate.

Huy-Ngo commented:
> Because you terminate before playback finishes. Quickest hack is to while `src.playing: pass`.

That doesn't change the result tho. Edit: by manually putting the string into the initializer, I found out that white noise is the only thing that works, since it's independent of time.

> Please elaborate.

When I set whatever to the parameter `--waveform`, even its default value `sine`, it gives this [error](https://gist.github.com/Huy-Ngo/0d7b2c106903807b05f75d0e238f5cce). Edit: That's because I set `nargs=1`, which turns the parameters into a list. Removing that solve the problem.

Huy-Ngo committed:
Fix the generator functions for impulse and white noise

Huy-Ngo committed:
Fix argument parsing problems

McSinyx commented:
The formula I gave you was wrong, so it produces an inaudible frequency (also the default 44 kHz you gave is only audible to whales I think).  Here's the working patch (I used numpy to make use of vector operations of scipy.signal):
```diff
diff --git a/examples/palace-tonegen.py b/examples/palace-tonegen.py
index af86786..053afda 100644
--- a/examples/palace-tonegen.py
+++ b/examples/palace-tonegen.py
@@ -19,27 +19,28 @@
 
 from argparse import Action, ArgumentParser
 from array import array
-from math import sin
+from functools import partial
+from math import isclose, sin, pi
+from random import random
 from typing import Callable, Dict, Tuple
 
 from palace import Buffer, Context, BaseDecoder, Device
-
-from numpy.random import random
+from numpy import arange, float32, ndarray, sin, vectorize
 from scipy.signal import sawtooth, square, unit_impulse
 
-WAVEFORMS: Dict[str, Callable[[float], float]] = {
+WAVEFORMS: Dict[str, Callable[[ndarray], ndarray]] = {
     'sine': sin,
     'square': square,
     'sawtooth': sawtooth,
-    'triangle': lambda time: sawtooth(time, 0.5),
-    'impulse': unit_impulse,
-    'white-noise': random}
+    'triangle': partial(sawtooth, width=0.5),
+    'impulse': lambda frames: unit_impulse(len(frames)),
+    'white-noise': vectorize(lambda time: random())}
 
 
 class ToneGenerator(BaseDecoder):
-    def __init__(self, waveform: str, duration: float, frequency: int):
-        self.func = lambda frame: WAVEFORMS[waveform](
-            frame/self.frequency/frequency)
+    def __init__(self, waveform: str, duration: float, frequency: float):
+        self.func = lambda frames: WAVEFORMS[waveform](
+            frames * frequency * pi * 2 / self.frequency)
         self.duration = duration
         self.start = 0
 
@@ -64,9 +65,9 @@ class ToneGenerator(BaseDecoder):
 
     def read(self, count: int) -> bytes:
         stop = min(self.start + count, self.length)
-        data = array('f', map(self.func, range(self.start, stop)))
+        data = self.func(arange(self.start, stop))
         self.start = stop
-        return data.tobytes()
+        return data.astype(float32).tobytes()
 
 
 class TypePrinter(Action):
@@ -75,23 +76,23 @@ class TypePrinter(Action):
         parser.exit()
 
 
-def play(device: str, waveform: str, duration: float, frequency: int) -> None:
+def play(device: str, waveform: str, duration: float, frequency: float) -> None:
     with Device(device) as dev, Context(dev):
         dec = ToneGenerator(waveform, duration, frequency)
         with Buffer.from_decoder(dec, 'tonegen') as buf, buf.play() as src:
-            _ = src
+            while src.playing: pass
 
 
 if __name__ == '__main__':
     parser = ArgumentParser()
     parser.add_argument('-t', '--types', nargs=0, action=TypePrinter,
                         help='print available waveform types in this example')
-    parser.add_argument('-w', '--waveform', default='sine', nargs=1,
-                        help='waveform to be generated')
+    parser.add_argument('-w', '--waveform', default='sine',
+                        help='waveform to be generated, default to sine')
     parser.add_argument('-d', '--device', default='', help='device name')
     parser.add_argument('-l', '--duration', default=5.0, type=float,
-                        help='duration, in second')
-    parser.add_argument('-f', '--frequency', default=44100, type=int,
-                        help='frequency for the wave')
+                        help='duration in second, default to 5')
+    parser.add_argument('-f', '--frequency', default=440.0, type=float,
+                        help='frequency in hertz, default to 440')
     args = parser.parse_args()
     play(args.device, args.waveform, args.duration, args.frequency)
```

Note that impulse it's just a click.

Huy-Ngo commented:
> The formula I gave you was wrong

I figured that out, since square wave only generated a load of 1. Sawtooth and triangle gave two "puff" sounds. I changed the formula given above to `frame/self.freqency * frequency` and it seems to give the right signal (high pitch sound). (No, apparently not, since I'm not supposed to hear 44kHz sound)

McSinyx commented:
See the diff above, it's welly tested in comparison with `speaker-test` already.

Huy-Ngo committed:
Fix the generator function in ToneGenerator

Huy-Ngo commented:
We don't need to operate on the whole array here, doing with each element is also fine. I've just pushed one that has the sound, tested with sounds within human audible range.

McSinyx commented:
> We don't need to operate on the whole array here, doing with each element is also fine.

That's what I thought, until I realized how slow it its compared to the vectorized version.  This has more to do with the overhead SciPy creates to support vectorization:
```console
$ time python3 examples/palace-tonegen.py

real	0m5.499s
user	0m5.441s
sys	0m0.056s
$ time python3 examples/palace-tonegen.py -w sine

real	0m5.448s
user	0m5.375s
sys	0m0.060s
$ time python3 examples/palace-tonegen.py -w square

real	0m10.839s
user	0m10.833s
sys	0m0.048s
```

After applying the following patch
```diff
diff --git a/examples/palace-tonegen.py b/examples/palace-tonegen.py
index 617c0d0..d670679 100644
--- a/examples/palace-tonegen.py
+++ b/examples/palace-tonegen.py
@@ -18,22 +18,23 @@
 # along with palace.  If not, see <https://www.gnu.org/licenses/>.
 
 from argparse import Action, ArgumentParser
-from array import array
-from math import pi, sin
+from functools import partial
+from math import pi
+from operator import not_
 from random import random
 from typing import Callable, Dict, Tuple
 
+from numpy import arange, float32, ndarray, sin, vectorize
 from palace import Buffer, Context, BaseDecoder, Device
-
 from scipy.signal import sawtooth, square
 
-WAVEFORMS: Dict[str, Callable[[float], float]] = {
+WAVEFORMS: Dict[str, Callable[[ndarray], ndarray]] = {
     'sine': sin,
     'square': square,
     'sawtooth': sawtooth,
-    'triangle': lambda time: sawtooth(time, 0.5),
-    'impulse': lambda time: 1 if time == 0 else 0,
-    'white-noise': lambda time: random() * 2 - 1}
+    'triangle': partial(sawtooth, width=0.5),
+    'impulse': vectorize(not_),
+    'white-noise': vectorize(lambda time: random()*2 - 1)}
 
 
 class ToneGenerator(BaseDecoder):
@@ -64,9 +65,9 @@ class ToneGenerator(BaseDecoder):
 
     def read(self, count: int) -> bytes:
         stop = min(self.start + count, self.length)
-        data = array('f', map(self.func, range(self.start, stop)))
+        data = self.func(arange(self.start, stop))
         self.start = stop
-        return data.tobytes()
+        return data.astype(float32).tobytes()
 
 
 class TypePrinter(Action):
```
the performance hugely improved:
```console
$ time python3 examples/palace-tonegen.py -w square

real	0m5.428s
user	0m5.342s
sys	0m0.080s
```

BTW this example is still not executable.

Huy-Ngo committed:
Optimize by using numpy array

Huy-Ngo committed:
Fix linting

Huy-Ngo committed:
Make the example files executable

Huy-Ngo commented:
Alright, things should be fine now.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,101 @@
+from typing import Callable, Dict, Tuple
+
+from palace import Buffer, Context, BaseDecoder, Device
+
```
From the example point of view, all numpy, palace and scipy are third party and should be grouped together.

Huy-Ngo replied:
Well, didn't I? I put them into three groups: builtin libraries, palace, and numpy/scipy.

Huy-Ngo replied:
Oh I see what you mean.

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,101 @@
+        dec = ToneGenerator(waveform, duration, frequency)
+        with Buffer.from_decoder(dec, 'tonegen') as buf, buf.play() as src:
+            while src.playing:
+                pass
```
This is a bit CPU intensive.  Try to sleep or something.

McSinyx replied:
Oh no you need to provide an argument to `sleep`!  Either a small value inside of the loop or Just `sleep(duration)` and forget about `src.playing`.  I'll nit it myself then (along with another few stuff).

Huy-Ngo replied:
Whoops, it ran so I forgot that 😕

McSinyx commented on examples/palace-tonegen.py:
```diff
@@ -0,0 +1,101 @@
+    parser = ArgumentParser()
+    parser.add_argument('-t', '--types', nargs=0, action=TypePrinter,
+                        help='print available waveform types in this example')
+    parser.add_argument('-w', '--waveform', default='white-noise', type=str,
+                        help='waveform to be generated,'
+                             'default to white-noise')
```
Personally I don't feel that white noise is a sane default value.  Maybe I just don't like chaos *that* much.

Huy-Ngo replied:
Lol, I set that as default just like the old TV does.

McSinyx reviewed (changes requested):
Just a few nits and then we're good to go!

Huy-Ngo committed:
Resolve request: sleep and change default waveform

Huy-Ngo committed:
Resolve request: group inputs together
